library/ssl_tls.c

```diff
@@ -1032,20 +1035,41 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
 #endif
 
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
-    ret = mbedtls_cipher_setup_psa( &transform->cipher_ctx_enc,
-                                    cipher_info, taglen );
-    if( ret != 0 && ret != MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE )
+
+    /* Only use PSA-based ciphers for TLS-1.2.
+     * That's relevant at least for TLS-1.0, where
+     * we assume that mbedtls_cipher_crypt() updates
+     * the structure field for the IV, which the PSA-based
+     * implementation currently doesn't. */
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
```

Comment:
Suggestion: The code to setup the context is getting non-trivial and is quite duplicated between setting up the encrypt and decrypt contexts. Should we extract it to a helper function?

Reply:
In general, yes, but concretely I'm inclined to leave things as they are. Firstly, `mbedtls_ssl_derive_keys()` is very large (we should split it at some point), so introducing a static function would move it quite far away from the point where it is used, which impedes readability. , Secondly, I prefer to make it obvious that the code is the same as before when `MBEDTLS_USE_PSA_CRYPTO` isn't set. 

I am not fixed to the current version and will change it if you insist, but otherwise I'd like to keep it as is for now.

Reply:
Good point. Let's keep it that way for now, then.